### PR TITLE
Bug 1863606 - Add smoke test generation to experiment integration test suite

### PR DESCRIPTION
**Pull request by jrbenny35**
[JIRA Ticket](https://mozilla-hub.atlassian.net/browse/EXP-4006)
### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1863606

### CODE REVIEW

File: fenix/app/src/androidTest/java/org/mozilla/fenix/experimentintegration/generate_smoke_tests.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    path = Path("../ui")
+    files = sorted([x for x in path.iterdir() if x.is_file()])
+    locations = []
+    file_name = None
+    test_names = []
+
+    for name in files:
+        if tests_name in name.name:
+            file_name = name
+            break
```
Comment (Trinaa) [inline fenix/app/src/androidTest/java/org/mozilla/fenix/experimentintegration/generate_smoke_tests.py:18]: Maybe `next()` could be used here?

I think it's technically possible for `file_name` to remain `None` after this loop. Perhaps add a check before opening the file?

File: fenix/app/src/androidTest/java/org/mozilla/fenix/experimentintegration/generate_smoke_tests.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+    with open(file_name, "r") as file:
+        code = file.read().split(" ")
+        code = [item for item in code if item != ""]
+
+        for count, item in enumerate(code):
+            if "class" in item:
+                locations.append(count)
+            if "@SmokeTest" in item:
+                locations.append(count)
```
Comment (Trinaa) [inline fenix/app/src/androidTest/java/org/mozilla/fenix/experimentintegration/generate_smoke_tests.py:28]: Just checking... is it intended that `count` would be appended twice should `"class"` and `"@SmokeTest"` be found in item?

At first glance I thought this, but maybe not...
```
if "class" in item or "@SmokeTest" in item:
    locations.append(count)
```

File: fenix/app/src/androidTest/java/org/mozilla/fenix/experimentintegration/gradlewbuild.py [hunk 3]
```diff
@@ -16,12 +16,15 @@ class GradlewBuild(object):
...
+    def test(self, identifier, smoke=None):
         # self.adbrun.launch()
 
         # Change path accordingly to go to root folder to run gradlew
         os.chdir("../../../../../../../..")
-        cmd = f"adb shell am instrument -w -e class org.mozilla.fenix.experimentintegration.{identifier} org.mozilla.fenix.debug.test/androidx.test.runner.AndroidJUnitRunner"
+        if smoke:
+            cmd = f"adb shell am instrument -w -e class org.mozilla.fenix.ui.{identifier} org.mozilla.fenix.debug.test/androidx.test.runner.AndroidJUnitRunner"    
+        else:
+            cmd = f"adb shell am instrument -w -e class org.mozilla.fenix.experimentintegration.{identifier} org.mozilla.fenix.debug.test/androidx.test.runner.AndroidJUnitRunner"
```
Comment (Trinaa) [inline fenix/app/src/androidTest/java/org/mozilla/fenix/experimentintegration/gradlewbuild.py:27]: Maybe this (with a better parameter name then `cmd_value` 😆)?

```suggestion
        cmd_value = "ui" if smoke else "experimentintegration"
        cmd = f"adb shell am instrument -w -e class org.mozilla.fenix.{cmd_value}.{identifier} org.mozilla.fenix.debug.test/androidx.test.runner.AndroidJUnitRunner"
```

**Review by Trinaa — COMMENTED** [review]
Hi @jrbenny35 👋🏻,  
I looked things over and left some comments, I didn't execute the code and I don't have solid context, so please take or leave things 😸

File: fenix/app/src/androidTest/java/org/mozilla/fenix/experimentintegration/generate_smoke_tests.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        for item in tests:
+            file.writelines(f"{item}")
+
+
+if __name__ == "__main__":
+    test_modules = None
+    create_test_file()
+    with open("variables.yaml", "r") as file:
+        test_modules = yaml.safe_load(file)
+    for item in test_modules["smoke_tests"]:
```
[the remark this answers is not in the document]
Comment (jrbenny35) [reply fenix/app/src/androidTest/java/org/mozilla/fenix/experimentintegration/generate_smoke_tests.py:82]: I love my ```KeyError```s 😵‍💫

**Review by jjSDET — APPROVED** [review]
LGTM; Also, kudos on the self discoverable test generation. A++

**Pull request merged**